### PR TITLE
Add Test Case - improvements to CTE feature

### LINKED ISSUE #87: improvements to CTE feature
When creating CTEs, we have to carry all of them, it'd be better that the user only indicated one and we figure out the rest. Example:

```sql
%%sql --save positive_x
SELECT * FROM numbers
WHRE x > 0
````

```sql
%%sql --save positive_x_and_y --with positive_x
SELECT * FROM positive_x
WHRE y > 0
````

If I want to use `positive_x_and_y`, I need:

```sql
%%sql  --with positive_x_and_y --with positive_x
SELECT * FROM numbers
WHRE x > 0
````

But having `--with positive_x_and_y` should be enough, and JupySQL should figure out that it also needs `positive_x`

### PR DESCRIPTION
**Pull request by tonykploomber**
## Describe your changes

Add test case to cover `--save` and `--with` query usage

Per [this request](https://github.com/ploomber/jupysql/issues/87#issuecomment-1437645060)

## Issue ticket number and link
Closes #87

## Checklist before requesting a review

- [ ] I have performed a self-review of my code
- [ ] I have added thorough [tests](https://github.com/ploomber/contributing/blob/main/CONTRIBUTING.md#testing) (when necessary).
- [ ] I have added the right documentation in the [docstring](https://github.com/ploomber/contributing/blob/main/CONTRIBUTING.md#documenting-changes-and-new-features) and [changelog](https://github.com/ploomber/contributing/blob/main/CONTRIBUTING.md#changelog) (when needed)

----
:books: Documentation preview :books:: https://jupysql--238.org.readthedocs.build/en/238/

### CODE REVIEW

**Review by edublancas — CHANGES_REQUESTED** [review]
we need to update the test, it should be three queries. the second one should use the first one and the third one should use the second one. but we need to test that the third one only needs to pass `--with second` and figure out that it also needs the first one (even if not explicitly passed)

**Review by neelasha23 — APPROVED** [review]

**Comment by tonykploomber** [comment]
> we need to update the test, it should be three queries. the second one should use the first one and the third one should use the second one. but we need to test that the third one only needs to pass `--with second` and figure out that it also needs the first one (even if not explicitly passed)

Updated test case in [f42f4ae](https://github.com/ploomber/jupysql/pull/238/commits/f42f4ae9778a76a32c26a57f4a8dbfd4577a36fe)

- Add third query

To construct the third query, we only need to provide `--with` second is enough

File: src/tests/test_magic.py [hunk 1]
```diff
@@ -603,3 +603,31 @@ def test_columns_with_dollar_sign(ip_empty):
     html = result.result._repr_html_()
 
     assert "$2" in html
+
+
+def test_save_with(ip):
+    # First Query
+    ip.run_cell(
+        "%sql --save shakespeare SELECT * FROM author WHERE last_name = 'Shakespeare'"
```
Comment (yafimvo) [inline src/tests/test_magic.py:611]: I think there are only 2 records in this table, so running `SELECT * FROM author WHERE last_name = 'Shakespeare'"` will always return 1 record, so the limit of 10 is not needed and we might miss some edge cases here (e.g: 2nd and 3rd queries return different results, None, etc...)

Maybe we can test it with more queries and verify we are getting the expected results for each. we can do it with `@pytest.mark.parametrize` and basically keep the same structure.

Also, I would add some error tests, like what happens when there is no such table, bad input, etc...

**Review by yafimvo — CHANGES_REQUESTED** [review]

File: src/tests/test_magic.py (hunk 1 above)
Comment (tonykploomber) [reply src/tests/test_magic.py:611]: I agree we should use more data to test it, fixed in [7d91bf8](https://github.com/ploomber/jupysql/pull/238/commits/7d91bf87fdeae784f9612d774480599b431d9b69)

For the error tests, let me know what you think we should cover more?

Comment (yafimvo) [reply src/tests/test_magic.py:611]: I think let's test the most common use cases: bad table inputs (for --save and --with) and bad queries. If anything else comes up in your mind go for it.

**Review by edublancas — APPROVED** [review]
